### PR TITLE
Refactor inputs/outputs and stack overflow handling at start/end

### PR DESCRIPTION
**Pull request by grjte**
This PR addresses #347 to remove the restriction on the number of inputs that can be provided and to remove the requirement that the stack overflow table be emptied before program execution ends.

- [x] enable providing more than 16 inputs
- [x] enable ending execution without emptying the stack overflow table
- [x] add new struct for program outputs (completed as part of the above), enabling more output flexibility in the future
- [x] add air tests for different possible numbers of inputs
- [x] add air tests for different possible final stack states (with/without overflow)
- [x] adjust boundary constraints for initial stack state and adds overflow constraint
- [x] adjust boundary constraints for final stack and adds overflow constraint
- [x] update existing tests to remove the extra `drop` instructions which are no longer needed
- [x] update docs & stdlib for new changes
- [x]  cleanup pass (any further docs updates and changes related to questions below, e.g. specification of num of outputs)
- [x] update CLI for new output handling

## Questions:

- the current change to outputs returns 16 elements from the top of the stack along with the rows on the stack overflow table (required for verifying the proof). However, we could all items on the stack as an array (in addition to extra info required for verifying the proof). We could also continue allowing specification of the number of outputs (currently removed)
- would we prefer to handle program outputs differently, especially regarding the verifier? For example, currently I've created a struct that can be provided to the verifier with the integer versions of program outputs, but maybe we'd rather do a vector u64 (for the stack) and a vector of tuples (for the overflow rows)
- should we rename `finalize_stack`?

### CODE REVIEW

**Comment by grjte** [comment]
@bobbinth it would be great to get your thoughts on the 2 questions I left in the PR description before I finish checking all the docs and finalize this for review

**Comment by bobbinth** [comment]
@grjte some (preliminary) thoughts from me:

1. I'd probably store the stack items in a vector in `ProgramOutputs`. We could optimize it to store the entire overflow table as a single field element - but that can be done in a separate PR. Besides the stack items values, what else do we need to store there? It seems to me that we should be able to reconstruct overflow table rows just from these values.
2. Regarding returning the specified number of outputs - I think we should do this. We could add another field to `ProgramOutputs` - e.g. `num_requested_outputs`, and then just have a method that returns a slice of such length.
3. I think providing a struct is better than providing two separate vectors as it will give us flexibility to optimize how we store data in this struct in the future. I also think it should be the same struct as is returned by the prover - unless there are good reasons to do two different stucts.
4. I'm fine with renaming `finzlize_stack` into something else - but don't have a name suggestion.

**Comment by grjte** [comment]
>     1. I'd probably store the stack items in a vector in `ProgramOutputs`. We could optimize it to store the entire overflow table as a single field element - but that can be done in a separate PR. Besides the stack items values, what else do we need to store there? It seems to me that we should be able to reconstruct overflow table rows just from these values.

We need to know the clock cycles when things were added to the table for the boundary constraint on the end of the stack overflow table's auxiliary column. (Currently I have the full row, which has clk and prev. We could get away without prev as long as things are ordered, but I wanted to check in on this first)

>     2. Regarding returning the specified number of outputs - I think we should do this. We could add another field to `ProgramOutputs` - e.g. `num_requested_outputs`, and then just have a method that returns a slice of such length.
>     3. I think providing a struct is better than providing two separate vectors as it will give us flexibility to optimize how we store data in this struct in the future. I also think it should be the same struct as is returned by the prover - unless there are good reasons to do two different stucts.
> 
>     4. I'm fine with renaming `finzlize_stack` into something else - but don't have a name suggestion.

Great, I'll update these items tomorrow morning & then finalize this PR for review. I don't have an alternate suggestion for `finalize_stack` yet either. I just thought the name might not align clearly with the use case anymore.

**Comment by bobbinth** [comment]
> We need to know the clock cycles when things were added to the table for the boundary constraint on the end of the stack overflow table's auxiliary column.

Ah yes! Makes sense. I'd still probably have a vector for stack values and then put clock cycles etc. into a separate field.

> I don't have an alternate suggestion for `finalize_stack` yet either.

Maybe `truncate_stack` or something like that?

**Comment by grjte** [comment]
@bobbinth one question regarding the CLI - all of the outputs are required in order for verification to succeed, but we may still want to allow people to specify a number of outputs. This would probably imply 2 output files - one for the user, and one to be read in by the system during verification. What are your thoughts on this? Also, would it be better to do this as a separate PR? (this would mean leaving the CLI broken) I don't mind doing it as part of this one, but I want to make sure we're agreed on the output format before I make those changes.

File: prover/src/lib.rs [hunk 1]
```diff
@@ -80,15 +78,22 @@ pub fn prove(
...
+    overflow_inputs: Vec<Felt>,
+    outputs: ProgramOutputs<Felt>,
 }
 
 impl ExecutionProver {
-    pub fn new(options: ProofOptions, num_stack_inputs: usize, num_stack_outputs: usize) -> Self {
+    pub fn new(
+        options: ProofOptions,
+        num_stack_inputs: usize,
+        overflow_inputs: Vec<Felt>,
```
Comment (bobbinth) [inline prover/src/lib.rs:89]: I wonder if we should just pass the full list of inputs here. Then, we can use these directly in construction of public inputs (instead of reading them from the trace).

We could also add a `debug_assert` in `get_pub_inputs()` to make sure values in the trace are consistent with inputs/outputs.

File: processor/src/lib.rs [hunk 2]
```diff
@@ -83,21 +83,21 @@ pub struct ChipletsTrace {
...
 
-/// Returns an execution trace resulting from executing the provided program against the provided
-/// inputs.
+/// Returns execution output and an execution trace resulting from executing the provided program
+/// against the provided inputs.
 pub fn execute(
     program: &Program,
     inputs: &ProgramInputs,
-) -> Result<ExecutionTrace, ExecutionError> {
+) -> Result<(ProgramOutputs<Felt>, ExecutionTrace), ExecutionError> {
```
Comment (bobbinth) [inline processor/src/lib.rs:91]: Would it complicate things much if we still just returned `ExecutionTrace` from here and modified the `ExecutionTrace` struct to contain (and return via a method) `ProgramOutputs`? If it doesn't complicate things too much, I'd prefer this approach.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! This is not yet a full review - but I left some comments inline about general structure. To summarize the main ones:

1. I'd prover not to make `ProgramOutputs` struct generic.
2. I wonder if it would be possible to keep the processor's `execute()` interface the same (i..e, return `ExecutionTrace`) and then the trace itself would be modified to store all required data to return inputs and outputs.

Regarding your question about number of outputs. I think we should still have it at CLI level, it it would be relevant only when outputs are printed to stdout. So, no need to two output files. If outputs are written to a file, we'd always write the full set of outputs (including overflow table addresses).

I'd prefer to fix the CLI as a part of this PR.

File: prover/src/lib.rs [hunk 3]
```diff
@@ -79,17 +71,53 @@ pub fn prove(
...
+            .take(self.stack_inputs.len())
+            .enumerate()
+        {
+            if *input != self.stack_inputs[i] {
+                valid = false;
+            }
         }
+
+        valid
+    }
```
Comment (bobbinth) [inline prover/src/lib.rs:105]: There probably should be a `break;` after `valid = false;` here and in the `validate_ouputs()` method below.

nit: since these return booleans, I'd prefer to name them something like `are_inputs_valid()` and `are_outputs_valid()`.

Also, curious why you chose to index into `stack_inputs` rather than doing something like:
```Rust
for (t_input, p_input) in trace.init_stack_state().iter().zip(self.stack_inputs()) {
```

File: processor/src/stack/aux_trace.rs [hunk 4]
```diff
@@ -64,4 +72,28 @@ impl AuxColumnBuilder<OverflowTableUpdate, OverflowTableRow> for AuxTraceBuilder
...
+
+    /// Returns the initial value in the auxiliary column.
+    fn init_column_value<E: FieldElement<BaseField = Felt>>(&self, row_values: &[E]) -> E {
+        let mut init_column_value = E::ONE;
+        // iterate through the elements in the initial table
+        for (_, hint) in &self.overflow_hints[..self.num_init_rows] {
+            // no rows should have been removed from the table before execution begins.
+            if let OverflowTableUpdate::RowInserted(row) = hint {
+                init_column_value *= row_values[*row as usize];
+            }
```
Comment (bobbinth) [inline processor/src/stack/aux_trace.rs:84]: If I understood things correctly, all hints should be `RowInserted`. If so, might be good to add a `debug_assert` (maybe in the `else` branch) to make sure this is always the case?

File: air/src/stack/mod.rs [hunk 5]
```diff
@@ -1,3 +1,146 @@
...
+/// The number of auxiliary assertions.
+pub const NUM_AUX_ASSERTIONS: usize = 2;
+
+// BOUNDARY CONSTRAINTS
+// ================================================================================================
+
+// --- MAIN TRACE ---------------------------------------------------------------------------------
+
+/// Returns the stack's boundary assertions for the main trace at the first step.
+pub fn get_assertions_first_step(result: &mut Vec<Assertion<Felt>>, stack_inputs: &[Felt]) {
```
Comment (bobbinth) [inline air/src/stack/mod.rs:29]: I think we are missing a couple of assertions here (they were probably missing before this PR too):

1. Stack column $b_0$ should be set to the depth of the stack.
2. Stack column $b_1$ should be set to the address of the first row in the overflow table.

File: processor/src/stack/trace.rs [hunk 6]
```diff
@@ -21,10 +21,10 @@ pub struct StackTrace {
 impl StackTrace {
     // CONSTRUCTOR
     // --------------------------------------------------------------------------------------------
-    /// Returns a [StackTrace] instantiated with empty vectors for all columns.
-    pub fn new(inputs: &ProgramInputs, init_trace_capacity: usize) -> Self {
+    /// Returns a [StackTrace] instantiated with the provided input values. When fewer than
+    /// `MIN_STACK_DEPTH` inputs are provided, the rest of the stack top elements are set to ZERO.
+    pub fn new(init_values: &[Felt], init_trace_capacity: usize) -> Self {
```
Comment (bobbinth) [inline processor/src/stack/trace.rs:26]: I think we need to make a couple of updates to the construction:

1. Column $b_0$ should be initialized to the actual stack depth. It seems like right now it is always initialized to $16$. We should probably also add a test to catch a situation where the actual stack depth does not match what is recorded in $b_0$.
2. Column $b_1$ should be initialized to the address of the first row in the overflow table. Previously, it didn't need to be done since the overflow table was always empty and the value in this column was initialized to zero.

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! I left a few comments inline.

One other thing: we should probably create an issue to update design docs in the future to make sure the explanation of how the overflow table works reflects the latest updates.

File: core/src/outputs.rs [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+    /// Returns an iterator of the elements that were on the top of the stack at the end of
+    /// execution.
+    pub fn stack_top_iter(&self) -> Box<dyn Iterator<Item = Felt> + '_> {
+        Box::new(
+            self.stack
+                .iter()
+                .take(MIN_STACK_DEPTH)
+                .map(|v| Felt::new(*v)),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (grjte) [reply core/src/outputs.rs:83]: There are a few places where we need an iterator over the stack top (as field elements). I was trying to avoid iterating over the whole thing twice every time (once to convert to Felt and once to use it). Is there a better way to return an iterator so that we just iterate once (converting to field elements and then continuing with the other iteration steps)?

File: prover/src/lib.rs (hunk 3 above)
Comment (grjte) [reply prover/src/lib.rs:105]: Yep, happy to add the break and change the names. The reason I didn't use zip is that there's no guarantee they're the same length. `stack_inputs` could be shorter or longer than `init_stack_state`. If I did it the way you proposed we'd have to have more length checks, so I thought it would be more complex and less readable.

File: processor/src/stack/aux_trace.rs (hunk 4 above)
Comment (grjte) [reply processor/src/stack/aux_trace.rs:84]: yes, this is correct. I'll add a `debug_assert`

File: air/src/stack/mod.rs (hunk 5 above)
Comment (grjte) [reply air/src/stack/mod.rs:29]: Yes, we're missing AIR constraints - I thought we were leaving stack AIR constraints for another PR. I only added the overflow aux column one because it's required for this PR. The others are only here because they were already defined (and also needed to be adjusted as a requirement of this PR).

Do you want these constraints added in this PR?

File: processor/src/stack/trace.rs (hunk 6 above)
Comment (grjte) [reply processor/src/stack/trace.rs:26]: Great point, thank you!

File: core/src/outputs.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bobbinth) [reply core/src/outputs.rs:83]: Not sure there is a better way. But we call this method only in a handful of places, and we iterate over only 16 elements. So, I wouldn't worry about performance implications. I'd rather have a bit simpler code.

File: processor/src/lib.rs [hunk 8]
```diff
@@ -155,13 +155,19 @@ impl Process {
...
             self.system.clk(),
             0,
             "a program has already been executed in this process"
         );
-        self.execute_code_block(program.root(), program.cb_table())
+        self.execute_code_block(program.root(), program.cb_table())?;
+
+        // return all values on the stack as program outputs
+        let stack = self.stack.get_values(None);
+        let overflow_addrs = self.stack.get_overflow_addrs();
```
[the remark this answers is not in the document]
Comment (grjte) [reply processor/src/lib.rs:168]: Sure, I'll update it.

File: core/src/outputs.rs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+    /// first).
+    pub fn stack_overflow_iter(&self) -> Box<dyn Iterator<Item = Felt> + '_> {
+        Box::new(
+            self.stack
+                .iter()
+                .skip(MIN_STACK_DEPTH)
+                .rev()
+                .map(|v| Felt::new(*v)),
+        )
+    }
```
[the remark this answers is not in the document]
Comment (grjte) [reply core/src/outputs.rs:96]: Yep, it makes sense to combine these. Regarding the iterators, I have the same comment/question as above about iterating twice

File: prover/src/lib.rs (hunk 3 above)
Comment (bobbinth) [reply prover/src/lib.rs:105]: I might be wrong, but I thought `zip` would produce elements only until both iterators return `Some(...)`. That is, as soon as one of the iterators stops, `zip`'ed iterator would stop as well.

File: air/src/stack/mod.rs (hunk 5 above)
Comment (bobbinth) [reply air/src/stack/mod.rs:29]: Since we are enforcing constraints for stack start/end, might as well enforce constraints for the bookkeeping columns too (unless it is a complicated thing to do).

File: prover/src/lib.rs (hunk 3 above)
Comment (grjte) [reply prover/src/lib.rs:105]: Ah you're right! I thought I'd read it the other way, but I clearly couldn't have, because the docs agree with you :) Yep, `zip` would be more convenient.

File: processor/src/stack/mod.rs [hunk 10]
```diff
@@ -68,11 +68,34 @@ impl Stack {
...
+        let (trace, overflow) = if init_values.len() > MIN_STACK_DEPTH {
+            let overflow = OverflowTable::new_with_inputs(
+                keep_overflow_trace,
+                &init_values[MIN_STACK_DEPTH..],
+            );
+            let trace = StackTrace::new(
+                &init_values[..MIN_STACK_DEPTH],
+                init_trace_capacity,
+                depth,
+                Felt::new(Felt::MODULUS - 1),
```
Comment (bobbinth) [inline processor/src/stack/mod.rs:83]: nit: I would have probably used `-ONE` instead of `Felt::new(Felt::MODULUS - 1)`. But we can change this later.

File: processor/src/stack/tests.rs [hunk 11]
```diff
@@ -11,17 +13,52 @@ fn initialize() {
...
+    let mut stack_inputs = [
+        1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19,
+    ];
+    let inputs = ProgramInputs::new(&stack_inputs, &[], vec![]).unwrap();
+    let stack = Stack::new(&inputs, 4, false);
+
+    // Prepare the expected results.
+    stack_inputs.reverse();
+    let expected_stack = build_stack(&stack_inputs[..MIN_STACK_DEPTH]);
+    let expected_helpers = [Felt::new(19), Felt::new(Felt::MODULUS - 1), Felt::ZERO];
```
Comment (bobbinth) [inline processor/src/stack/tests.rs:41]: This may not matter for the first row, but I think we should set the last helper column to 1/(b0 - 16). We can do it in a follow-up PR.

**Review by bobbinth — APPROVED** [review]
Looks good! Thank you! There are a couple of minor comments, but I'll merge as is and we can address them in the future PR.
